Adopt AdAttributionKit's private browsing verification protocol

achristensen07 opened the pull request:
#### 5d5bb2812b45a398e30816a86da84dbe50dd4948
<pre>
Adopt AdAttributionKit&apos;s private browsing verification protocol
<a href="https://bugs.webkit.org/show_bug.cgi?id=282629">https://bugs.webkit.org/show_bug.cgi?id=282629</a>
<a href="https://rdar.apple.com/137446845">rdar://137446845</a>

Reviewed by John Wilander.

We tell AdAttributionKit the session identifier of private browsing sessions if they are the source
of impression donations.  When the session is destroyed, we notify AdAttributionKit.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::didClose):
* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::notifyAdAttributionKitOfSessionTermination):
* Source/WebKit/NetworkProcess/cocoa/AppStoreDaemonSPI.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::~NetworkSessionCocoa):
(WebKit::NetworkSessionCocoa::notifyAdAttributionKitOfSessionTermination):
(WebKit::NetworkSessionCocoa::donateToSKAdNetwork):

Canonical link: <a href="https://commits.webkit.org/287354@main">https://commits.webkit.org/287354@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/245cf4e996fe22d9144de8398152b8117c492f64

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79110 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58146 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32487 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83748 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30320 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6415 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61910 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19829 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82177 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51958 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72106 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42214 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49309 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26222 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28687 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72233 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70425 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26643 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85135 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78325 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6431 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4450 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70150 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6594 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67977 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69398 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13449 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12252 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100648 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12240 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6383 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12195 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21964 "Found 2 new JSC stress test failures: wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-eager-jettison, wasm.yaml/wasm/stress/struct-new_default-small-members.js.wasm-no-cjit (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6336 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9793 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8128 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/AppStoreDaemonSPI.h:
```diff
@@ -60,4 +60,17 @@ typedef NS_ENUM(NSInteger, ASDInstallWebAttributionContext) {
+// FIXME: Move these to the !USE(APPLE_INTERNAL_SDK) section once rdar://137446922 is complete.
+#if HAVE(AD_ATTRIBUTION_KIT_PRIVATE_BROWSING)
+NS_ASSUME_NONNULL_BEGIN
+@interface ASDInstallWebAttributionService (Staging_for_137446922)
```
This "staging for" thing is news to me. What does it control or convey?

achristensen07 replied:
Once that radar is completed and included in an SDK we can move this to the section of the SPI header that is only for non-internal SDKs.  Until then we need a selector declaration when building with internal and non-internal SDKs.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -1519,7 +1519,23 @@ - (void)URLSession:(NSURLSession *)session webSocketTask:(NSURLSessionWebSocketT
-NetworkSessionCocoa::~NetworkSessionCocoa() = default;
+NetworkSessionCocoa::~NetworkSessionCocoa()
+{
+    notifyAdAttributionKitOfSessionTermination();
```
Is there any risk we get killed before we can complete this notification? I'm thinking of when we get jetsammed. Just trying to understand if this will be "leaky" or not.

For crashes, I guess all bets are off.

achristensen07 replied:
That is a problem that is not solved yet, but the common case of closing the browser hits this.

johnwilander commented on Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
```diff
@@ -2182,10 +2198,19 @@ void completeWithBlockedError()
-#endif
-#endif
+#endif // HAVE(ASD_INSTALL_WEB_ATTRIBUTION_SERVICE)
+#endif // HAVE(SKADNETWORK_v4)
```
Do we have other references to SKAdNetwork here, and this is just a clarification of something old? Or should we be using the AdAttributionKit name instead? (v4 sounds old to me so may be fine)

achristensen07 replied:
This is old.  We should come by later and clean it up.

johnwilander reviewed (approved):
Please see comments. Other than that, looks good.

achristensen07 committed:
Adopt AdAttributionKit's private browsing verification protocol
https://bugs.webkit.org/show_bug.cgi?id=282629
rdar://137446845

Reviewed by John Wilander.

We tell AdAttributionKit the session identifier of private browsing sessions if they are the source
of impression donations.  When the session is destroyed, we notify AdAttributionKit.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::didClose):
* Source/WebKit/NetworkProcess/NetworkSession.h:
(WebKit::NetworkSession::notifyAdAttributionKitOfSessionTermination):
* Source/WebKit/NetworkProcess/cocoa/AppStoreDaemonSPI.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::NetworkSessionCocoa::~NetworkSessionCocoa):
(WebKit::NetworkSessionCocoa::notifyAdAttributionKitOfSessionTermination):
(WebKit::NetworkSessionCocoa::donateToSKAdNetwork):

Canonical link: https://commits.webkit.org/287354@main
